### PR TITLE
Add What I Talk About When I Talk About Query Optimizer (Part 1): IR Design

### PR DESCRIPTION
**Pull request by Xuanwo**

### CODE REVIEW

**Comment by Xuanwo** [comment]
cc @leiysky for reivew

File: content/2024/02-exploring-sql-query-optimizer-part-1.md [hunk 1]
```diff
@@ -0,0 +1,954 @@
+---
+categories: Code
+date: 2024-01-29T01:00:00Z
+title: "Exploring SQL Query Optimizer (Part 1): IR Design"
```
Comment (leiysky) [inline content/2024/02-exploring-sql-query-optimizer-part-1.md:4]: The origin of the title is a Haruki Murakami's book **"What I Talk About When I Talk About Running"**, so I think we can translate the title into **"What I Talk About When I Talk About Query Optimizer"**.

Comment (Xuanwo) [reply content/2024/02-exploring-sql-query-optimizer-part-1.md:4]: Got it.

File: content/2024/02-what-i-talk-about-when-i-talk-about-query-optimizer-part-1.md [hunk 2]
```diff
@@ -0,0 +1,953 @@
...
+
+Most of the operator implementations went smoothly, but it can be seen that `Project` and `Aggregate` have been marked as todo. At this point, we will find that `Project` and `Aggregate` cannot directly generate their own attributes using `children attributes`. 
+
+Going back to relational algebra, the purpose of `Project` is to trim the shape of tuples or modify the name of attributes. This kind of SQL expression like `SELECT a + 1 AS b FROM t` cannot be expressed as a naive `Project`; as for `Aggregate`, it is not even present in basic relational algebra, it is an extension to relational algebra.
+
+The theory of relational algebra no longer exists!
+
+However, despite this, the project still needs to continue. We need to introduce some *village rules* to expand the definition of relational algebra. Here we provide the formal definitions of `Project` and `Aggregate` in `The IR`.
+
+![](project.svg)
```
Comment (leiysky) [inline content/2024/02-what-i-talk-about-when-i-talk-about-query-optimizer-part-1.md:470]: Ditto.

File: content/2024/02-what-i-talk-about-when-i-talk-about-query-optimizer-part-1.md [hunk 3]
```diff
@@ -0,0 +1,953 @@
...
+
+`Project` represents the attributes in relationship `R` as input, output a tuple consisting of `n` function mappings `f_1` to `f_n`.
+
+![](aggregate.svg)
```
Comment (leiysky) [inline content/2024/02-what-i-talk-about-when-i-talk-about-query-optimizer-part-1.md:474]: Ditto.

**Comment by leiysky** [comment]
Rest LGTM, thanks for your work very much!

File: content/2024/02-what-i-talk-about-when-i-talk-about-query-optimizer-part-1.md [hunk 4]
```diff
@@ -0,0 +1,953 @@
...
+- Data distribution 
+- Data partition
+
+Combining the properties of relational algebra, we can describe the differences between types of properties. 
+
+Assuming there are relations `R` and `S`: the relational property of `R` is `RP_R`, and the physical property is `PP_R`; the relational property of `S` is `RP_S`, and the physical property is `PP_S`.
+
+We can obtain:
+
+![](property.svg)
```
[the remark this answers is not in the document]
Comment (Xuanwo) [reply content/2024/02-what-i-talk-about-when-i-talk-about-query-optimizer-part-1.md:375]: Fixed.

File: content/2024/02-what-i-talk-about-when-i-talk-about-query-optimizer-part-1.md (hunk 3 above)
Comment (Xuanwo) [reply content/2024/02-what-i-talk-about-when-i-talk-about-query-optimizer-part-1.md:474]: Fixed.

**Review by leiysky — APPROVED** [review]

**Pull request merged**